txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,482 @@
+    pub fn find_dependent(
+        &self,
+        tx: ArcTx,
+        seen: &mut HashMap<TxId, ArcTx>,
+        txs: &HashMap<TxId, ArcTx>,
+    ) {
+        // for every input aggregate UtxoId and check if it is inside
+        let mut check = vec![tx.id()];
+        while let Some(parent_txhash) = check.pop() {
+            let mut is_new = false;
+            let mut parent_tx = None;
+            seen.entry(parent_txhash).or_insert_with(|| {
+                is_new = true;
+                let parent = txs.get(&parent_txhash).expect("To have tx in txpool");
+                parent_tx = Some(parent.clone());
+                parent.clone()
+            });
+            // for every input check if tx_id is inside seen. if not, check coins/contract map.
+            if let Some(parent_tx) = parent_tx {
+                for input in parent_tx.inputs() {
+                    // if found and depth is not zero add it to `check`.
+                    match input {
+                        Input::Coin { utxo_id, .. } => {
+                            let state = self
+                                .coins
+                                .get(utxo_id)
+                                .expect("to find coin inside spend tx");
+                            // if depth is not zero it means tx is inside txpool. Zero == db utxo
+                            if state.depth != 0 {
+                                check.push(*utxo_id.tx_id());
+                            }
+                        }
+                        Input::Contract { contract_id, .. } => {
+                            let state = self
+                                .contracts
+                                .get(contract_id)
+                                .expect("Expect to find contract in dependency");
+
+                            if state.depth != 0 {
+                                let origin = state
+                                    .origin
+                                    .as_ref()
+                                    .expect("contract origin to be present");
+                                check.push(*origin.tx_id());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn check_if_coin_input_can_spend_output(
+        output: &Output,
+        input: &Input,
+        is_output_filled: bool,
+    ) -> anyhow::Result<()> {
+        if let Input::Coin {
+            owner,
+            amount,
+            color,
+            ..
+        } = input
+        {
+            let i_owner = owner;
+            let i_amount = amount;
+            let i_color = color;
+            match output {
+                Output::Coin { to, amount, color } => {
+                    if to != i_owner {
+                        return Err(Error::NotInsertedIOWrongOwner.into());
+                    }
+                    if amount < i_amount {
```

Comment:
Is there any reason why we'd not enforce the amount to be the same? Seems unlikely people would burn their own tokens like this, but seems like it would be better to ensure the amount in == amount out.

Reply:
I was strict and used == first, but i relaxed restriction few commits back. will return too stricter option.